actioncable/test/connection/stream_test.rb

```diff
@@ -37,18 +37,21 @@ def on_error(message)
 
   [ EOFError, Errno::ECONNRESET ].each do |closed_exception|
     test "closes socket on #{closed_exception}" do
-      run_in_eventmachine do
-        connection = open_connection
+      # StringIO couldn't be registered with NIO (and we don't need this functionality for the test)
+      @server.event_loop.stub(:attach, nil) do
+        run_in_eventmachine do
+          connection = open_connection
 
-        # Internal hax = :(
-        client = connection.websocket.send(:websocket)
-        rack_hijack_io = client.instance_variable_get("@stream").instance_variable_get("@rack_hijack_io")
-        rack_hijack_io.stub(:write, proc { raise(closed_exception, "foo") }) do
-          assert_called(client, :client_gone) do
-            client.write("boo")
```

Comment:
I think there is a chance here to simplify the test removing some internals stubbing by making the IO object a real IO (and stubbing `write_nonblock` instead). 

```diff
diff --git a/actioncable/test/connection/stream_test.rb b/actioncable/test/connection/stream_test.rb
index 0f4576db40..03703380f1 100644
--- a/actioncable/test/connection/stream_test.rb
+++ b/actioncable/test/connection/stream_test.rb
@@ -38,12 +38,11 @@ def on_error(message)
   [ EOFError, Errno::ECONNRESET ].each do |closed_exception|
     test "closes socket on #{closed_exception}" do
       run_in_eventmachine do
-        connection = open_connection
+        rack_hijack_io = File.open(File::NULL, "w")
+        connection = open_connection(rack_hijack_io)
 
-        # Internal hax = :(
-        client = connection.websocket.send(:websocket)
-        rack_hijack_io = client.instance_variable_get("@stream").instance_variable_get("@rack_hijack_io")
-        rack_hijack_io.stub(:write, proc { raise(closed_exception, "foo") }) do
+       client = connection.websocket.send(:websocket)
+        rack_hijack_io.stub(:write_nonblock, proc { raise(closed_exception, "foo") }) do
           assert_called(client, :client_gone) do
             client.write("boo")
           end
@@ -54,11 +53,11 @@ def on_error(message)
   end
 
   private
-    def open_connection
+    def open_connection(rack_hijack_io)
       env = Rack::MockRequest.env_for "/test",
         "HTTP_CONNECTION" => "upgrade", "HTTP_UPGRADE" => "websocket",
         "HTTP_HOST" => "localhost", "HTTP_ORIGIN" => "http://rubyonrails.com"
-      env["rack.hijack"] = -> { env["rack.hijack_io"] = StringIO.new }
+      env["rack.hijack"] = -> { env["rack.hijack_io"] = rack_hijack_io }
 
       Connection.new(@server, env).tap do |connection|
         connection.process
```

Reply:
👍 Thanks!
Have to replace `:write` with `:write_nonblock` in the stub, but otherwise works like a charm

Reply:
Great! Yes, my diff was already using `:write_nonblock` too. I see some tests are failing but they look unrelated. I'll get this merged with the help of @jeremy. Thanks!